src/precompiles/native.rs

```diff
@@ -65,190 +66,295 @@ pub fn is_valid_account_id(account_id: &[u8]) -> bool {
+            .expect(ERR_TARGET_TOKEN_NOT_FOUND),
+    )
+    .unwrap()
+}
+
+pub struct ExitToNear<S>(PhantomData<S>); //TransferEthToNear
 
-impl Precompile for ExitToNear {
+impl<S: AuroraState> Precompile<S> for ExitToNear<S> {
     fn required_gas(_input: &[u8]) -> Result<u64, ExitError> {
         Ok(costs::EXIT_TO_NEAR_GAS)
     }
 
     #[cfg(not(feature = "contract"))]
-    fn run(input: &[u8], target_gas: u64, context: &Context) -> PrecompileResult {
+    fn run(
+        input: &[u8],
+        target_gas: u64,
+        context: &Context,
+        _state: &mut S,
+        _is_static: bool,
+    ) -> PrecompileResult {
         if Self::required_gas(input)? > target_gas {
             return Err(ExitError::OutOfGas);
         }
 
-        Ok((ExitSucceed::Returned, Vec::new(), 0))
+        Ok(PrecompileOutput {
+            exit_status: ExitSucceed::Returned,
+            output: Vec::new(),
+            cost: 0,
+            logs: Vec::new(),
+        })
     }
 
     #[cfg(feature = "contract")]
-    fn run(input: &[u8], target_gas: u64, context: &Context) -> PrecompileResult {
+    fn run(
+        input: &[u8],
+        target_gas: u64,
+        context: &Context,
+        state: &mut S,
+        is_static: bool,
+    ) -> PrecompileResult {
         if Self::required_gas(input)? > target_gas {
             return Err(ExitError::OutOfGas);
         }
 
-        let (nep141_address, args) = if context.apparent_value != U256::from(0) {
-            // ETH transfer
-            //
-            // Input slice format:
-            //      recipient_account_id (bytes) - the NEAR recipient account which will receive NEP-141 ETH tokens
-
-            if is_valid_account_id(input) {
-                (
-                    String::from_utf8(crate::sdk::current_account_id()).unwrap(),
-                    crate::prelude::format!(
-                        r#"{{"receiver_id": "{}", "amount": "{}", "memo": null}}"#,
-                        String::from_utf8(input.to_vec()).unwrap(),
-                        context.apparent_value.as_u128()
-                    ),
-                )
-            } else {
-                return Err(ExitError::Other(Cow::from(
-                    "ERR_INVALID_RECEIVER_ACCOUNT_ID",
-                )));
+        if is_static {
+            return Err(ExitError::Other(Cow::from("ERR_INVALID_IN_STATIC")));
+        }
+
+        // First byte of the input is a flag, selecting the behavior to be triggered:
+        //      0x0 -> Eth transfer
+        //      0x1 -> Erc20 transfer
+        let mut input_mut = input;
+        let flag = input[0];
+        input_mut = &input_mut[1..];
+
+        let (nep141_address, args) = match flag {
+            0x0 => {
+                // ETH transfer
+                //
+                // Input slice format:
+                //      recipient_account_id (bytes) - the NEAR recipient account which will receive NEP-141 ETH tokens
+
+                if is_valid_account_id(input) {
```

Comment:
```suggestion
                if is_valid_account_id(input_mut) {
```

Reply:
Nice catch! to avoid this error, I replaced all `input_mut` with `input`, and shadowed previos `input` with a mutable version of itself.